# Assignment 1
## use this template!

This is the Notebook template that you would need for submitting your Assignment 1. Follow the order of this Notebook in reporting. If that is not followed, the assignment is not passed, hence resulting in 0 points.

The objective for this assignment is to build classification model to classify different types of network attacks in the CICIDS2017 dataset.
 - See Assignment 1 in Moodle for more detailed information

In general, when you get going with the assignment, it is good to note, that your codes would not (most probably) be such that you would run them on one machine and hence that your code would be just one file. Rather, right at the beginning, start planning together a skeleton for your code that serves as starting point for every group member. Then you divide among your group, which classification algorithm is handled by whom. Possibly by defining then more precisely which exact hyperparameters and pipeline steps you would utilize for that particular model. Read through this Notebook template and I believe you start to get the idea of how you could start approaching this assignment!

Below, I have provided topics which need to be answered below it. Do not change the order.

---

## 1. The Group 
- List briefly the group name, group member names and the study groups for the group members

---

## 2. Contributions for the group members
Notice, basically everybody need to be able to do every step required. Of course, it may be that the group will work on the skeleton for the trainings together. Remember to say that if so. But if somebody is clearly not knowledgeable on, say, PCA and has not done anything with that, the group must make that clear here. In a way, that will become clear below with the results but that needs to be said here also.
- Explain the  of each group member.

| Group member name | Contibution |
| :--- | :--- |
| First Name, Last Name | Example: Did study all the steps for the pre-processing for example, was conducting the trainings and testing for kNN,<br> Was considering hyper parameter k and PCA, ran experiment w/o SMOTE also|
|  |  |
|  |  |
|  |  |
|  |  |
|  |  |




---

---

## 3. Use of AI
- Explain, how did you utilize AI to help you with the assignment?
- Each member does reporting they status individually

| Group member name | Contibution |
| :--- | :--- |
| First Name, Last Name | Example: I did not use AI at all|
|  |  |
|  |  |
|  |  |
|  |  |
|  |  |

---

## 4. Dataset, Splitting the dataset
- How is the data sampled for the training, validation and testing dataset?

- How did you make sure everyone is using the same data?



In [1]:
# all libraries here

import numpy as np
import pandas as pd
import sklearn
import imblearn
import xgboost
import matplotlib
import seaborn

---

In [2]:
df_original = pd.read_csv("cicids2017_cleaned.csv", low_memory=False)
df_100 = df_original.sample(n=100_000, random_state=42)

In [3]:
#### 20.4. MK: this was only to compare the sample dataset to the original one

counts_original = df_original["Attack Type"].value_counts()
percentages_original = df_original["Attack Type"].value_counts(normalize=True)

counts_100 = df_100["Attack Type"].value_counts()
percentages_100 = df_100["Attack Type"].value_counts(normalize=True)

comparison = pd.DataFrame({
    "Row count original" : counts_original, 
    "Count-% original": percentages_original.round(5)*100,
    "Row count sample file" : counts_100, 
    "Count-% sample file": percentages_100.round(5)*100})

print(comparison)

                Row count original  Count-% original  Row count sample file  \
Attack Type                                                                   
Normal Traffic             2095057            83.112                  83081   
DoS                         193745             7.686                   7673   
DDoS                        128014             5.078                   5115   
Port Scanning                90694             3.598                   3574   
Brute Force                   9150             0.363                    377   
Web Attacks                   2143             0.085                     92   
Bots                          1948             0.077                     88   

                Count-% sample file  
Attack Type                          
Normal Traffic               83.081  
DoS                           7.673  
DDoS                          5.115  
Port Scanning                 3.574  
Brute Force                   0.377  
Web Attacks           

In [5]:
df_100.to_csv("cleaned_100k.csv", index=False)

In [6]:
# The dataset and dataframe to be used

df = pd.read_csv("cleaned_100k.csv", low_memory=False)

The original dataset contained over 2.3 million rows. The original CSV file was sampled with Pandas to 100 000 items with random state of 42. If the original dataframe based on cicids2017_cleaned.csv would be read in, with random state = 42 the sample dataframe would be exactly the same.

 In dataset cicids2017_cleaned.csv there are 7 unique attack types, it is a paramount to have each type represented in the sample file. This way only the train-test model can learn about more rare attack type classes. For performance and computational purposes, randomly chosen 100 000 rows from the original set were chosen.

The comparison between original dataset and sample file is represented in this table:
| Attack Type     | Row Count (Original) | Percent % (Original) | Row Count (Sample) | Percent % (Sample) |
|-----------------|-----------------------|------------------------|---------------------|----------------------|
| Normal Traffic  | 2 095 057             | 83.112                 | 83 081              | 83.081               |
| DoS             |   193 745             | 7.686                  | 7 673               | 7.673                |
| DDoS            |   128 014             | 5.078                  | 5 115               | 5.115                |
| Port Scanning   |    90 694             | 3.598                  | 3 574               | 3.574                |
| Brute Force     |     9 150             | 0.363                  |   377               | 0.377                |
| Web Attacks     |     2 143             | 0.085                  |    92               | 0.092                |
| Bots            |     1 948             | 0.077                  |    88               | 0.088                |

The percentages of each attack type are well represented in the dataset, as it mimics the original dataset's propotions.

The dataframe was exported into CSV file named cleaned_100k.csv. This CSV is also used as a basis for the assignment's dataframe, which confirms, that the assignment is based solely on the sampled dataset and dataframe.

## 5. Label variable

- What type of classification you are dealing with. Binary or Multi class? Make it clear such that what will follow in this Report makes sense to me.


---

## 6. Feature variables
- How did you select which features you include in the trainings?


---

## 7. SMOTE
- Explain shortly how SMOTE took place in your tests. Report your overall findings for SMOTE already here even though you may or may not say something on SMOTE later in the Results section. I say this like this because, the results section may become populated with lots of results already otherwise, so this might be your opportunity to keep that section simpler.


---

## 8. Scaling, Polyfeatures, PCA, LDA, Kernel PCA


- Similar to SMOTE above. You will most probably run lots of tests on your own separate Notebooks different from this group level submission Notebook.
- Again, the results section may become populated with lots of results already otherwise, so this might be your opportunity to keep that section simpler.
- Still say conclusions shortly, what worked, what not
- Paste here a code sample/snippet showing the main steps, their tested orders and settings you used for these. Yes, the code snipped may not be functional on this notebook, as this is supposed to be just snippet illustrating your working. Adding all, would probably make the document messy!
- So explain main effect these step had for your trainings. Did something work or not, so basically when did you got benefit from those.
- Which parameters did provide the best results in case of which algorithm


---

## 9. Training setups
- Provide snippets of code that would through Pipeline and GripSearchCV variables make it clear which hyperparameters did you cover, what where the order of steps for standardization, feature dimensionality reduction etc. Ok, maybe bit overlapping with the part 8. above but at least the classification algorithms were not covered there. 
- I wrote snippets(!) in plural form as for different members of the group, you may have needed to make adjustments to those variables depending on, for example, the classification algorithm that you personally have been working with.
- However, do not exaggerate. Keep the snippets concise and short, the point is to show and convince me that you have done what you were supposed to and you know what you were doing
- As mentioned at the end of the Assignment 1 description in Moodle, in addition to this Notebook, zip your personal working Notebooks were you show more functional code. However, I check these only if needed for some reason. So basically, the idea is to write here things so that I would not need to check those. So it is about the clarity of the submission in a way.
 

---

## 10. Classification Results and Best model/models
- Again, make sure you use same data for training, validation and testing.
- Collect your results then together here
- Provide accuracies and confusion matrices. Maybe selfevident, but the results need to be comparable for different algorithms. So present them that way also.
- You can also consider precision, recall and F1-score. At least Recall may be interresting to observe when comparing the performance with or without SMOTE
- What is then the best model/models, point that out
- What is then the setup (preprocessing steps, hyperparameter) that produced that best model(s)?


---

Finally, make sure for the last time that this documentation of yours would make sense for somebody who might continue your work with extended data later on.
